docs/src/ref/modeling.md

```diff
@@ -97,31 +97,46 @@ val::Bool = @trace(bernoulli(0.5), :z)
+
+For example, the sample space of [`Bernoulli`](@ref) is `Bool` and its support is either `{true}`, `{false}`, or `{true, false}`. The sample space of [`Normal`](@ref) is `Real` and its support is the set of all values on the real line. The sample space of [`Beta`](@ref) is `Real` and its support is the set of values in the interval (0, 1).
+
+Gen's built in modeling languages require that a address is associated with a fixed sample space. For example, it is not permitted to use a `Bernoulli` distribution to sample at addresss `:a` in one execution, and a `Normal` distribution to sample at address `:a` in a different execution, because their sample spaces differ (`Bool` vs `Real`):
 ```julia
 @gen function foo()
-    n = @trace(categorical([0.5, 0.5]), :n) + 1
-    @trace(categorical(ones(n) / n), :x)
+    if @trace(bernoulli(0.5), :branch)
+        @trace(bernoulli(0.5), :x)
+    else
+        @trace(normal(0, 1), :x)
+    end
 end
 ```
-The support of the random choice with address `:x` is either the set ``\{1, 2\}`` or ``\{1, 2, 3\}``.
-Therefore, this random choice does satisfy our condition above.
-This would cause an error with the following, in which the `:n` address is modified, which could result in a change to the domain of the `:x` variable:
-```julia
-tr, _ = generate(foo, (), choicemap((:n, 2), (:x, 3)))
-tr, _ = mh(tr, select(:n))
-```
-We can modify the address to satisfy the condition by including the domain in the address:
+
+A generative function can be **disciplined** or not. In a disciplined generative function, the support of random choices at each address must be fixed. That is, there exists a set S that is a subset of the sample space such that for all executions of the generative function, if an address appears in the execution, its support must be S. Violating this discipline will cause NaNs, errors, or undefined behavior in certain inference programs. However, there are many inference programs where violating this discipline does not result in undefined behavior and is convenient.
+
+If the support of a random choice needs to change, a disciplined generative function can represent this by using a different address for each distinct value of the support. For example, consider the following generative function:
 ```julia
 @gen function foo()
     n = @trace(categorical([0.5, 0.5]), :n) + 1
-    @trace(categorical(ones(n) / n), (:x, n))
+    @trace(categorical(ones(n) / n), :x)
 end
 ```
+The support of the random choice with address `:x` is either the set ``\{1, 2\}`` or ``\{1, 2, 3\}``.
+Therefore, this random choice does not have constant support, and the generative function `foo` is not 'disciplined'.
+Specifically, this could result in undefined behavior for the following inference program:
+```julia
+tr, _ = importance_resampling(foo, (), choicemap((:x, 3)))
```

Comment:
Thanks for writing up these docs @marcoct! They provide a lot of clarity. My only issue is maybe with this example of possible undefined behavior. With the changes made in this pull request, I tested this exact example, and `importance_resampling` correctly does rejection sampling and returns only traces with with `tr[:n] = 2` (which is the only way that `tr[:x]` could have been `3`:
```julia
julia> traces = [importance_resampling(foo, (), choicemap((:x, 3)), 100)[1] for i=1:1000];
julia> all(tr -> tr[:n] == 2, traces)
true
```
I'm confident that `importance_resampling` will correctly reject `-Inf` traces in other cases as well, based on how the current implementation is written. Specifically, it will never enter this branch of the code for a `-Inf` trace:
```
# Line 102 of src/inference/importance.jl
        if bernoulli(exp(log_weight - log_total_weight))
            model_trace = cand_model_trace
        end
```
So it seems overly conservative to say that `importance_resampling` will result in undefined behavior. Unless you're referring to the fact that very (very) rarely, `importance_resampling` will return a `-Inf` trace, because it never manages to propose a trace with non-zero probability.

Reply:
> Unless you're referring to the fact that very (very) rarely, importance_resampling will return a -Inf trace, because it never manages to propose a trace with non-zero probability.

@ztangent Yes, that's what I'm referring to.

Reply:
And it's not clear that it should return anything in that case -- I'm not sure what the behavior is right now, but I think maybe it should just crash. We could define the behavior to just return a random trace if it receives all -Inf weights but that seems not obvious, and the same sort of decision would need to be made in various other inference primitives (e.g. primitives SMC) as well as by user inference programs (which may not be as careful as the built-in primitives in their handling of -Infs and NaNs).

Reply:
I am a fan of removing the original suggested fix though, where we include the support in the trace address:
```julia
@gen function foo()
    n = @trace(categorical([0.5, 0.5]), :n) + 1
    @trace(categorical(ones(n) / n), (:x, n))
end
```
The reason is that if you define, `foo` this way, and then do:
```julia
tr, _ = importance_resampling(foo, (), choicemap((:x, 3) => 3), 100)
```
what ends up happening is that the constraint `(:x, 3) => 3` ends up being totally ignored when :n is sampled to be 1. Effectively, this means that `:n=1` ends up being sampled with relative probability `0.5`, while `:n=2` ends up being sampled with relative probability `0.5 * 1/3` (because all traces with `:n=2` are constrained to satisfy `(:x, 3)=3`, and end up being down-weighted by 1/3).

IMO this is very unexpected behavior! It certainly was to me when I first tried to do something like this in Gen. And I don't think this sort of behavior is well documented either. What I *would* have expected is that all traces with `:n=1` be rejected, because they fail to ensure that `(:x, 3)` exists in the trace.  But Gen currently doesn't have logic that checks this*. So to me, the undisciplined version of `foo` actually exhibits more well-defined behavior than the disciplined version, because `importance_resampling` correctly does rejection sampling (with the changes in this PR).

*And I understand that adding the logic to check this might mess up some other things in Gen, because sometimes Gen allows unsampled addresses to be in the choicemap, as the current implementation of `regenerate` allows for.

Reply:
> IMO this is very unexpected behavior! It certainly was to me when I first tried to do something like this in Gen. And I don't think this sort of behavior is well documented either.

Yes. This is the behavior defined incredibly tersely in the current GFI documentation (in the subtle definition of the behavior of q), but it could be much better documented, at both the GFI level and the inference library level (like what importance resampling does).  But I think this is a question for a different issue (https://github.com/probcomp/Gen/issues/208). But briefly,  I think that importance resampling should probably crash if not all the constraints are visited, but that there should be a flag to enable or disable this check (Gen's current approach to not always performing dynamic checks of this form is to allow for improved performance.).  I think that the GFI should keep its behavior of ignoring irrelevant constraints (or have an option for being more stringent), and that inference primitives should have an option for enabling dynamic checks for this sort of thing -- this is the approach taken by the various MCMC primitives -- e.g. you can enable dynamic checks that your observations aren't being over-written.

Reply:
> And it's not clear that it should return anything in that case -- I'm not sure what the behavior is right now, but I think maybe it should just crash. We could define the behavior to just return a random trace if it receives all -Inf weights but that seems not obvious, and the same sort of decision would need to be made in various other inference primitives (e.g. primitives SMC) as well as by user inference programs (which may not be as careful as the built-in primitives in their handling of -Infs and NaNs).

So the behavior right now (with this PR) is that it'll return a trace that is inconsistent with the observations, with weight `-Inf`:
```julia
@gen function foo()
    n = @trace(categorical([1.0, 0.0]), :n) + 1
    @trace(categorical(ones(n) / n), :x)
end
```
```julia
julia> tr, w = importance_resampling(foo, (), choicemap((:x, 3)), 10);
julia> tr[:n], w
1, -Inf
```
I think one possibility is to catch `-Inf` and throw an error. Another possibility is to just do a `@warn`ing? I feel like one reason why you might want to avoid throwing an error is that external code that calls `importance_sampling` or other inference algorithms might want to detect when `-Inf` traces are returned then automatically increase the number of samples without having to do a `try/catch` block. But anyway, none of this has to go into this PR, just thought I'd confirm what the current behavior is and have it documented here.

Reply:
@ztangent:

> So it seems overly conservative to say that importance_resampling will result in undefined behavior.

Aha, but (in my opinion) undefined is not about what the code will do, it's about whether the behavior is specified!  So if the spec says behavior for a certain kind of input is undefined, then it is truly undefined, even if a particular implementation always handles that input in a knowable way.